### PR TITLE
endpoint: skip Envoy incremental updates if no Envoy redirects

### PR DESCRIPTION
**Pull request by squeed**
This should significantly reduce the amount of time it takes to roll our incremental policy updates.

**Edit:** This PR was reverted and re-factored, see https://github.com/cilium/cilium/pull/31775

### CODE REVIEW

**Comment by squeed** [comment]
@jrajahalme / @sayboras Does this seem reasonable?

**Review by jrajahalme — APPROVED** [review]
This seems safe to do, as the altered function (`ApplyPolicyMapChanges()`) is only called from endpointmanager's `UpdatePolicyMaps()`, which is called from the FQDN and IPCache code paths.

File: pkg/endpoint/bpf.go [hunk 1]
```diff
@@ -1161,8 +1161,18 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
...
 	}
 
-	// Ignoring the revertFunc; keep all successful changes even if some fail.
-	err, _ = e.updateNetworkPolicy(proxyWaitGroup)
+	// Only update Envoy if there are envoy redirects.
+	// This is safe to do here, since a PolicyMapChange cannot
+	// cause an envoy redirect to appear or disappear. It only allows for
+	// incremental updates. Thus, we don't need to worry about stale
+	// policy not being cleaned up.
+	if e.desiredPolicy.L4Policy.HasEnvoyRedirect() {
```
Comment (jrajahalme) [inline pkg/endpoint/bpf.go]: add `|| e.isIngress` to the condition, as Ingress endpoint also needs to enforce L3/4 policy.

**Review by jrajahalme — CHANGES_REQUESTED** [review]

File: pkg/endpoint/bpf.go (hunk 1 above)
Comment (sayboras) [reply pkg/endpoint/bpf.go]: should we add one more condition for proxylib `e.desiredPolicy.L4Policy.HasProxylibRedirect()` as well ?

Comment (jrajahalme) [reply pkg/endpoint/bpf.go]: Proxylib is covered by Envoy check already, as it is also implemented in Envoy:
```
pkg/policy/l4.go:       redirectTypeProxylib redirectTypes = 1<<iota | redirectTypeEnvoy
```

**Review by sayboras — APPROVED** [review]
Thanks :heavy_check_mark:

**Comment by squeed** [comment]
@sayboras @jrajahalme I updated this slightly; we now skip Envoy if there are no redirects *or* network policy did not change.

File: pkg/endpoint/bpf.go [hunk 2]
```diff
@@ -1236,16 +1247,17 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 	}
 
 	if errors > 0 {
-		return fmt.Errorf("updating desired PolicyMap state failed")
+		return true, fmt.Errorf("updating desired PolicyMap state failed")
```
Comment (sayboras) [inline pkg/endpoint/bpf.go:1250]: (github didn't allow to comment on un-changed code, so I just add the comment here)

Does the above block guarded `e.visibilityPolicy != nil`  change policy map as well? This comment is not a blocker as visibility annotation feature is considered as deprecated.

Comment (squeed) [reply pkg/endpoint/bpf.go:1250]: > Does the above block guarded e.visibilityPolicy != nil change policy map as well?

Great question! (and thanks for asking it). Yes, `e.desiredPolicy.GetPolicyMap().AddVisibilityKeys(e, redirectPort, visMeta, changes)` adds new entries to `adds`. So, we have to check `len(adds)` at the end. (I had a version of this code where I got this wrong, heh).

**Review by sayboras — APPROVED** [review]
Still lgtm :heavy_check_mark:

File: pkg/endpoint/bpf.go [hunk 3]
```diff
@@ -1162,20 +1162,31 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
...
 
-	err := e.applyPolicyMapChanges()
+	changed, err := e.applyPolicyMapChanges()
 	if err != nil {
 		return err
 	}
 
-	// Ignoring the revertFunc; keep all successful changes even if some fail.
-	err, _ = e.updateNetworkPolicy(proxyWaitGroup)
+	// Only update Envoy if there are envoy redirects andthere were queued incremental changes.
```
Comment (jrajahalme) [inline pkg/endpoint/bpf.go]: typo: "andthere" -> "and there"

**Review by jrajahalme — APPROVED** [review]
Looks good, may fix typo in comment if need to respin  for something else.

**Pull request merged**